### ISSUE TITLE
Allow for attribution of all data (per-field sourcing)

### ISSUE THREAD

**Issue by evdb** [mysociety/popit#26]
Ensure that this suggestion is implemented:

> Make sure you include sources i.e. make it fundamental in the user interface that people put the URL and/or a text description of where they got data from.
> 
> This is vital for provenance and for later data updating. And most importantly, for the user interface to add credibility to sites made using the data.

**Comment by zarino** [comment]
Hi @kaerumy – If you had a way to record sources for each individual change, would you expect it to be optional, or required?

Also, if users were asked for a source *only once*, when they finally press the "Save changes" button, rather than individually per input field – would that be ok for your use case? If it wouldn't, could you explain why?

**Comment by zarino** [comment]
Just to add some context, YourNextMP (which is a user interface built on top of the PopIt API) has already made its own decisions about both of these points.

YourNextMP *requires* you to provide a source whenever you make any changes.

But it *doesn't* ask per individual field. Instead, on the candidate editing page (which is closest to the default PopIt person/organisation page) there's a single "Source" field next to the "Save" button at the bottom of the page:

![screen shot 2015-01-19 at 16 44 53](https://cloud.githubusercontent.com/assets/739624/5804038/a1f7f5a4-9ffa-11e4-8a13-c673896474f4.png)

This means YourNextMP can't track individual sources for each individual change, but it *can* record a source for each value changed at the same time (what I'd call an "editing session", or in version control might be called a "check-in" or a "commit").

Looking at the diffs on YourNextMP it appears most people only change one thing when they edit a candidate—maybe adding an email address or a twitter profile, or adding a membership—so recording one source per editing session, rather than per field, doesn't make much of a difference. But it *does* make it a lot simpler for users doing the editing.

The question is, @kaerumy, does Sinar Project need more granularity than that?

**Comment by kaerumy** [comment]
Hi @zarino, in short we're building a single centralized database of thousands of people & organizations from various verified sources, that will be used in by different applications as well as a research tool for investigative journalists for years to come. 

- each profile is built up from dozens of sources, often specific to one field. It's really important to know exactly where you got this from, especially if it could implicate someone to a corruption issue or even disqualify them for an election nomination. Technically you could have standard format for references to fields, and sources in a change log for the Sources field, but you would need to parse it for a few more of the following use cases
- Other users & applications could be using just a few fields. such as posts held. They are not interested in pulling in and digging through all sources, just displaying the source for these specific fields so others can quickly verify. eg. this person held post of CEO of ABC in 1992 (source: X url) One donor wanted to make sure that on website that contains multiple sources, that those sources for the field (eg. picture) are clearly displayed, to make sure that it's not randomly pulled off the Internet. 
- Editors/reviewers need to easily find/review source for specific field is for specific field. A person may hold post for a public company (source A), while holding a political post at same period (source B), before this he held a lower political post (source C), and at this specific time period, he was received gov award and title which adds another prefix/suffix to name (from source D).  For each of these fields,  a person could be holding up to 200 posts (from some research on beneficial ownership). That's a lot of sources, mixed in with changes for Twitter, FB etc. if in single unified change log.

**Comment by zarino** [comment]
Thanks for the feedback @kaerumy. You didn't quite answer my questions, but can I assume that your answers would be..?

* Yes, Sinar Project would expect source attribution to be a required step, performed by everyone who wants to submit new data.
* And no, citing a single source at the moment of saving is not good enough – Sinar would require individual source inputs for each changed field.

Is that accurate?

**Comment by kaerumy** [comment]
Yes correct on both.

**Comment by zarino** [comment]
For mySociety peeps – here's a run-down of my work on this so far:

https://docs.google.com/a/mysociety.org/presentation/d/1msmugkIGh3i1v25gYoq4DNULwyhcd7mGHtPy05u0FwQ/edit

**Comment by akuckartz** [comment]
This is about provenance and could use PROV-O (http://www.w3.org/TR/prov-o/).

**Comment by akuckartz** [comment]
PROV Primer: http://www.w3.org/TR/2013/NOTE-prov-primer-20130430/

**Comment by chrismytton** [comment]
Just a quick update on where we're at with this: We haven't forgotten about this ticket! We've done some work towards implementing it but it became clear along the way that the implementation is too coupled to the UI.

For the time being we've put this ticket on hold while we focus on decoupling the API and the UI, which is being tracked in #837.

Once we've made some progress with the API and UI decoupling then I want to come back to this ticket and hopefully the implementation will be much cleaner/clearer.

**Comment by kaerumy** [comment]
We've implemented this feature in our replacement version of Popit here https://github.com/Sinar/popit_ng
